### PR TITLE
Merge GLSL and HLSL fragment shader generators into one generic

### PR DESCRIPTION
**Pull request by hrydgard**
More refactoring. Finally we'll be able to make fragment shader changes in one single place. 

It has gotten a little bit more complex, but really not that bad.

Vertex shader generators coming up next.

### CODE REVIEW

File: GPU/Common/FragmentShaderGenerator.cpp [hunk 1]
```diff
@@ -283,22 +445,30 @@ bool GenerateFragmentShaderGLSL(const FShaderID &id, char *buffer, const GLSLSha
...
+		WRITE(p, "  PS_OUT outfragment;\n");		
+	} else if (compat.shaderLanguage == HLSL_D3D9) {
+		WRITE(p, "vec4 main( PS_IN In ) : COLOR {\n");
+	} else {
+		WRITE(p, "void main() {\n");
+	}
 	if (isModeClear) {
 		// Clear mode does not allow any fancy shading.
-		WRITE(p, "  vec4 v = v_color0;\n");
+		WRITE(p, "  vec4 v = %sv_color0;\n", compat.inPrefix);
```
Comment (aliaspider) [inline GPU/Common/FragmentShaderGenerator.cpp:458]: wouldn't it be simpler to define those directly under the `struct PS_IN` definition ?

Comment (hrydgard) [reply GPU/Common/FragmentShaderGenerator.cpp:458]: Ah, yeah, that would also be an option. Cleaner indeed, slightly more work for the shader optimizer but should be negligible.

Comment (hrydgard) [reply GPU/Common/FragmentShaderGenerator.cpp:458]: Oh right you mean the preprocessor define method, not declaring new variables in main. Yes actually that was the first thing I, but couldn't get it to work - some weird interaction with HLSL's preprocessor, I dunno. I'm gonna try again.

Comment (hrydgard) [reply GPU/Common/FragmentShaderGenerator.cpp:458]: Yeah, still not working - maybe you have an explanation for this?

```
   1:  #define vec2 float2
   2:  #define vec3 float3
   3:  #define vec4 float4
   4:  #define uvec3 uint3
   5:  #define ivec3 int3
   6:  #define splat3(x) float3(x, x, x)
   7:  #define mix lerp
   8:  #define mod(x, y) fmod(x, y)
   9:  #define DISCARD clip(-1)
  10:  #define DISCARD_BELOW(x) clip(x)
  11:  sampler tex : register(s0);
  12:  vec4 u_texclamp : register(c8);
  13:  vec4 u_alphacolorref : register(c1);
  14:  vec4 u_alphacolormask : register(c2);
  15:  vec3 roundAndScaleTo255v(float3 x) { return floor(x * 255.0f + 0.5f); }
  16:  struct PS_IN {
  17:    vec3 v_texcoord: TEXCOORD0;
  18:    vec4 v_color0: COLOR0;
  19:  };
  20:  #define v_color0 In.v_color0
  21:  #define v_color1 In.v_color1
  22:  #define v_texcoord In.v_texcoord
  23:  #define v_fogdepth In.v_fogdepth
  24:  vec4 main( PS_IN In ) : COLOR {
  25:    vec4 v = v_color0;
  26:    v.a = 0.0;
  27:    v.rgb = splat3(1.0) - v.rgb;
  28:    return v;
  29:  }
C:\dev\ppsspp\unittest\Shader@0x0000020389623FB0(25,12-19): error X3004: undeclared identifier 'v_color0'

```

Comment (aliaspider) [reply GPU/Common/FragmentShaderGenerator.cpp:458]: oh that's weird -.- maybe it didn't like the recursion in the macro or something ...

Comment (hrydgard) [reply GPU/Common/FragmentShaderGenerator.cpp:458]: I'm gonna do it with variable assignment instead, can't be bothered digging deeper into that..

Comment (aliaspider) [reply GPU/Common/FragmentShaderGenerator.cpp]: yeah it was the recursion, it works like this : 
```
struct PS_IN {
  vec3 texcoord: TEXCOORD0;
  vec4 color0: COLOR0;
};
#define v_color0 In.color0
#define v_color1 In.color1
#define v_texcoord In.texcoord
#define v_fogdepth In.fogdepth
```

Comment (hrydgard) [reply GPU/Common/FragmentShaderGenerator.cpp]: Scary with recursive defines, heh. I'll stick with the variables for now, works fine.

Comment (aliaspider) [reply GPU/Common/FragmentShaderGenerator.cpp]: it should have just ignored the recursion like any other C preprocessor, but no :/

**Pull request merged**